gno.land/pkg/gnoland/app.go

```diff
@@ -169,74 +181,122 @@ func NewApp(
-	return func(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
-		txResponses := []abci.ResponseDeliverTx{}
-
-		if req.AppState != nil {
-			// Get genesis state
-			genState := req.AppState.(GnoGenesisState)
-
-			// Parse and set genesis state balances
-			for _, bal := range genState.Balances {
-				acc := acctKpr.NewAccountWithAddress(ctx, bal.Address)
-				acctKpr.SetAccount(ctx, acc)
-				err := bankKpr.SetCoins(ctx, bal.Address, bal.Amount)
-				if err != nil {
-					panic(err)
-				}
-			}
+// InitChainerConfig keeps the configuration for the InitChainer.
+// [NewAppWithOptions] will set [InitChainerConfig.InitChainer] as its InitChainer
+// function.
+type InitChainerConfig struct {
+	// Handles the results of each genesis transaction.
+	GenesisTxResultHandler
+
+	// Standard library directory.
+	StdlibDir string
+	// Whether to keep a record of the DB operations to load standard libraries,
+	// so they can be quickly replicated on additional genesis executions.
+	// This should be used for integration testing, where InitChainer will be
+	// called several times.
+	CacheStdlibLoad bool
+
+	// These fields are passed directly by NewAppWithOptions, and should not be
+	// configurable by end-users.
+	baseApp *sdk.BaseApp
+	vmKpr   vm.VMKeeperI
+	acctKpr auth.AccountKeeperI
+	bankKpr bank.BankKeeperI
+}
+
+// InitChainer is the function that can be used as a [sdk.InitChainer].
+func (cfg InitChainerConfig) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
+	start := time.Now()
+	ctx.Logger().Debug("InitChainer: started")
+
+	{
+		// load standard libraries
+		// we write it to store right after loading the libs so that they are
+		// available when loading genesis txs.
+		// cache-wrapping is necessary for non-validator nodes; in the tm2 BaseApp,
+		// this is done in BaseApp.cacheTxContext; so we replicate it here.
+		ms := ctx.MultiStore()
+		msCache := ms.MultiCacheWrap()
+
+		stdlibCtx := cfg.vmKpr.MakeGnoTransactionStore(ctx)
+		stdlibCtx = stdlibCtx.WithMultiStore(msCache)
+		if cfg.CacheStdlibLoad {
+			cfg.vmKpr.LoadStdlibCached(stdlibCtx, cfg.StdlibDir)
+		} else {
+			cfg.vmKpr.LoadStdlib(stdlibCtx, cfg.StdlibDir)
+		}
+		cfg.vmKpr.CommitGnoTransactionStore(stdlibCtx)
+
+		msCache.MultiWrite()
+	}
+
+	ctx.Logger().Debug("InitChainer: standard libraries loaded",
+		"elapsed", time.Since(start))
 
-			// Run genesis txs
-			for _, tx := range genState.Txs {
-				res := baseApp.Deliver(tx)
-				if res.IsErr() {
-					ctx.Logger().Error(
-						"Unable to deliver genesis tx",
-						"log", res.Log,
-						"error", res.Error,
-						"gas-used", res.GasUsed,
-					)
-				}
-
-				txResponses = append(txResponses, abci.ResponseDeliverTx{
-					ResponseBase: res.ResponseBase,
-					GasWanted:    res.GasWanted,
-					GasUsed:      res.GasUsed,
-				})
-
-				resHandler(ctx, tx, res)
+	txResponses := []abci.ResponseDeliverTx{}
+
+	if req.AppState != nil {
```

Comment:
when is the case that AppState can be nil? Should we just error if that is the case?

Reply:
I've found it useful mostly when writing related tests (see app_test.go; both `testInitChainerLoadStdlib`, where we pass an empty abci.RequestInitChain, and `TestNewApp`). Avoids us creating dummy app states in those occasions.

I've made it log a warning if it's nil. Do you think it's OK? Do you think there's cases where it could happen in production and we should error instead?

Reply:
Yeah, if it is something that shouldn't happen in production, I would expect to do a sanity check and return an error if nil. If not we might be hiding some weird behaviors down the execution path.